### ISSUE TITLE
twisted.mail.smtp.sendmail should return a cancellable Deferred

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6572]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| kaizhang reported|
|-|-|
|Trac ID|trac#6572|
|Type|enhancement|
|Created|2013-06-14 05:03:12Z|
|Branch|https://github.com/twisted/twisted/tree/sendmail-deferred-cancellation-6572|

When a user call the `cancel` method of the Deferred returned by `twisted.mail.smtp.sendmail`, it should tell the `twisted.mail.smtp.SMTPSenderFactory` not to retry and disconnect the connection immediately.

Attachments:

* [6572-0.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a6e/a6ec4d22c66c3115c43d97dd45e90419e7374ef0/f46e1218bfcc309f7af6d4c70d71a77ce2f9ab67.patch) (5350 bytes) - added by kaizhang on 2013-06-14 11:10:11Z - Add cancellation support to `twisted.mail.smtp.sendmail`. Add test for it.
* [6572-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a6e/a6ec4d22c66c3115c43d97dd45e90419e7374ef0/083896d38763cfde824c65477ebd4514d10ebe0a.patch) (3021 bytes) - added by kaizhang on 2013-06-14 12:00:00Z - Add cancellation support to `twisted.mail.smtp.sendmail`. Write test code with `twisted.test.proto_helpers.MemoryReactor`.
* [6572-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a6e/a6ec4d22c66c3115c43d97dd45e90419e7374ef0/19ef58f04ca29dff2f823151498f8a36880eda38.patch) (6399 bytes) - added by kaizhang on 2013-06-17 07:20:54Z - Revise the patch according to the comments.

<details><summary>Searchable metadata</summary>

```
trac-id__6572 6572
type__enhancement enhancement
reporter__kaizhang kaizhang
priority__normal normal
milestone__ 
branch__branches_sendmail_deferred_cancellation_6572 branches/sendmail-deferred-cancellation-6572
branch_author__kaizhang kaizhang
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1371186192000000 1371186192000000
changetime__1373730229000000 1373730229000000
version__None None
owner__kaizhang kaizhang
cc__ashfall
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|kaizhang commented|
|-|-|

I added the cancellation support to `twisted.mail.smtp.sendmail`. But there is a known issue with the test code. If the smtphost passed to `twisted.mail.smtp.sendmail` is a name instead of IP address. The test will raises the DirtyReactorAggregateError, which suggests the DelayedCall - ThreadedResolver._cleanup - is neither called nor cancelled. I thought the connector would take care of this when disconnect. So I don't know if this is a bug. Please tell me if I did something wrong.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to kaizhang|
|-|-|

Thanks for your work on this!

 1. In the case where the connection to the server has already been made, you should be using `abortConnection()` to close it, but `connector.disconnect()` will use `loseConnection()`.
 1. The assertion about `factory.sendFinished` seems unnecessary.
 1. You should probably have an assertion that the default `_reactor` argument is the global reactor (you can use `inspect.getcallargs` to do that).
 1. The docstring for `sendmail` should be updated to document that the `Deferred` returned is cancellable, and explain what cancelling does.
 1. The docstring for the test should say `C{Deferred}` or maybe `L{defer.Deferred}` rather than `deferred`.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|kaizhang commented|
|-|-|

Replying to [itamar](#note_3):
> Thanks for your work on this!
> 
>  1. In the case where the connection to the server has already been made, you should be using `abortConnection()` to close it, but `connector.disconnect()` will use `loseConnection()`.
>  1. The assertion about `factory.sendFinished` seems unnecessary.
>  1. You should probably have an assertion that the default `_reactor` argument is the global reactor (you can use `inspect.getcallargs` to do that).
>  1. The docstring for `sendmail` should be updated to document that the `Deferred` returned is cancellable, and explain what cancelling does.
>  1. The docstring for the test should say `C{Deferred}` or maybe `L{defer.Deferred}` rather than `deferred`.

I have revised my patch according to the comments. There are two things worth noting:

1. I use `inspect.getargspec` to assert that the default `_reactor` argument is the global reactor. Because `inspect.getcallargs` is new in version 2.7 and `inspect.getargspec` is not.
2. Since `twisted.test.proto_helpers.StringTransport` doesn't have `abortConnection` method, I copied a version of `StringTransport` that supports `abortConnection` from `twisted.web.test.test_newclient`. Maybe we should add `abortConection` to `twisted.test.proto_helpers.StringTransport` later?

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kaizhang removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_7"></a>|@ashfall commented|
|-|-|

(In [38860]) Branching to 'sendmail-deferred-cancellation-6572'

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_8"></a>|@ashfall set owner to @ashfall<br>@ashfall set status to `assigned`|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/sendmail-deferred-cancellation-6572)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

Rather than copying the `StringTransport` class, I suggest just importing it from the other test module.

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_10"></a>|@ashfall set owner to kaizhang<br>@ashfall set status to `new`|
|-|-|

Thanks for working on this!

1. twisted/mail/test/test_smtp.py:

  1. test methods don't have to document their return type.
  2. `test_cancelBeforeConnectionMade` returns a Deferred. But you only need to do that if you need the test to run the global reactor. Since you're using MemoryReactor that is unnecessary. But it also means you should use `self.failureResultOf(d)` to extract the Deferred and assert its result immediately, rather than relying on assertFailure (see [the trial howto](http://twistedmatrix.com/documents/current/core/howto/trial.html)).
  3. Same for `test_cancelAfterConnectionMade`.
  4. Missing whitespace on line line 1713.
  5. Please use L{defer.Deferred} instead of C{Deferred}
  6. Is the line `connector = reactor.connectors[0]` in `test_cancelAfterConnectionMade` necessary?

2. twisted/mail/smtp.py

  1. Replace variable "argh" with "result" ("argh" is an exclamation of distress or annoyance).
  2. "p" is a bad name for an attribute. "currentProtocol" might be better, or come up with some other more informative name.
  3. The line `del self.p` in `removeProtocol` is unnecessary.

3. Missing news file.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|kaizhang removed owner|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=%2Fbranches%2Fsendmail-deferred-cancellation-6572)

I revised the patch according to the comments of itamar and ashfall. The pyflakes complains about "redefinition of unused 'StringIO' from line 44" in `twisted.mail.test.test_smtp`. I think it's a misreporting.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kaizhang set owner to kaizhang|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|kaizhang removed owner|
|-|-|

The private argument `_reactor` of `twisted.mail.smtp.sendmail` is changed to public argument `reactor`.

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/sendmail-deferred-cancellation-6572)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra commented|
|-|-|

twisted/mail/test/test_smtp.py imports `StringTransport` from `twisted.web.test.test_newclient`, making twisted.mail (its tests) dependant on twisted.web, which doesn't seem right. Why did you change the original import?

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|kaizhang commented|
|-|-|

Replying to [thijs](#note_14):
> twisted/mail/test/test_smtp.py imports `StringTransport` from `twisted.web.test.test_newclient`, making twisted.mail (its tests) dependant on twisted.web, which doesn't seem right. Why did you change the original import?

To test the cancellation, we need a `StringTransport` that support `abortConnection`. The original one doesn't support that. Itarmar and I discussed it and decided that we should file a ticket to add the `abortConnection` to the original `StringTransport`. Before that, Itarmar suggested just importing it from the other test module. You can see his comment in reply 9.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_16"></a>|@itamarst set owner to kaizhang|
|-|-|

Thanks for all your work!

 1. Switch back to copy/pasting AbortableStringTransport, but add a comment saying "This should be replaced by a common version in [#6530](https://github.com/twisted/twisted/issues/6530)." Not good, but yeah, we shouldn't have unnecessary cross-package dependencies.
 2. The docstring for sendmail() should say "it will stop retry" rather than "it will stop retrying".
 3. I think there's a bug in the logic for deciding what to do in cancellation. Consider the following scenario: connection succeeds, there's some problem, connection disconnects, new attempt at connection is made, operation is cancelled. The canceller will see that `currentProtocol` is set, but it's still the old protocol, and it will call `transport.abortConnection()` on it rather than stopping the new connection attempt. You should write a test for this edge case, and then once you've reproduced the problem fix the bug, or if you discover it's not an actual bug add a comment on the ticket explaining your reasoning.

Please fix the above and resubmit.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|kaizhang set status to `assigned`|
|-|-|

Replying to [itamar](#note_16):
>  2. The docstring for sendmail() should say "it will stop retry" rather than "it will stop retrying".

Hi, the docstring for sendmail() does say "it will stop retry" now. Should I change it to "it will stop retrying"?

**Comment by twisted-trac** [comment]
|[<img alt="ashfall's avatar" src="https://avatars.githubusercontent.com/u/1138873?s=50" width="50" height="50">](https://github.com/ashfall)<a name="note_18"></a>|@ashfall commented|
|-|-|

Replying to [kaizhang](#note_17):
> Replying to [itamar](#note_16):
> >  2. The docstring for sendmail() should say "it will stop retry" rather than "it will stop retrying".
> 
> Hi, the docstring for sendmail() does say "it will stop retry" now. Should I change it to "it will stop retrying"?

Yes please.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kaizhang removed owner<br>kaizhang set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|kaizhang commented|
|-|-|

Replying to [itamar](#note_16):

>  3. I think there's a bug in the logic for deciding what to do in cancellation. Consider the following scenario: connection succeeds, there's some problem, connection disconnects, new attempt at connection is made, operation is cancelled. The canceller will see that `currentProtocol` is set, but it's still the old protocol, and it will call `transport.abortConnection()` on it rather than stopping the new connection attempt. You should write a test for this edge case, and then once you've reproduced the problem fix the bug, or if you discover it's not an actual bug add a comment on the ticket explaining your reasoning.

I have reproduced the problem. I think it's a bug of `SMTPSenderFactory` processing client connection error after we added the `currentProtocol` attribute rather than a bug of cancellation. So I added tests for removing the current protocol when client connection is failed/lost and fixed the bug.

Also, since around half supported builders have been shutdown. I didn't force a build upon the new patch. I will add the build results once the baelnorn issue is addressed.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_21"></a>|@tomprince commented|
|-|-|

Looking at this code, I wondered if it might ever be the case that `.buildProtocol` has been calledc, but not `.makeConnection`. Looking at the code for TCP connections, it looks like those are always called synchronously, so that there is no opportunity for `.cancel` to be called between them.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_22"></a>|@tomprince set owner to kaizhang|
|-|-|

Thanks for your work on this.
 1. `currentProtocol`: Please add documentation for this in the class docstring.
 2. `sendFinished`:
   1. It appears that you are changing the meaning of this slightly. Please add documentation for this.
     - Originally, it being `1` (or `True`, this code probably predates `True`) meant that the message had been sent, or there has been an unrecoverable error.
     - Now, it may also mean that the sending should be cancelled.
   2. Also, you should set it to `True` instead of `1` in `cancel`.

Please resubmit after making the above changes.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_23"></a>|kaizhang removed owner|
|-|-|

I have revised my patch according to the comments.

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/sendmail-deferred-cancellation-6572)

Replying to [tom.prince](#note_22):

>    2. Also, you should set it to `True` instead of `1` in `cancel`.

The original implementation will compare `sendFinished` with 0 and `retries`(see `twisted.mail.smtp.SMTPSenderFactory._processConnectionError`), so I think it'd be better to set `sendFinished` to `1` in `cancel`.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_24"></a>|@tomprince set owner to kaizhang|
|-|-|

>> 2. Also, you should set it to `True` instead of `1` in `cancel`.

> The original implementation will compare `sendFinished` with 0 and `retries`(see `twisted.mail.smtp.SMTPSenderFactory._processConnectionError`), so I think it'd be better to set `sendFinished` to `1` in `cancel`.

I understand your desire to match the existing code, but in this case the existing code isn't an example of a style to follow.

The values you assign to `sendFinished` should be `True`/`False` and the documented type `bool`.

Old versions (before 2.2) of python didn't have `True` and `False`, and there is lots of code (like the code you are modifying it) that predates this. They thus had to use `1` and `0` for those values, and apparently, the clever
```python
if self.retries < self.sendFinished <= 0:
```
In new code, this would be better expressed as 
```
#!
if (self.retries < 0) and (not self.sendFinished):
```
You shouldn't change that in this ticket but, you shouldn't avoid using `True` and `False` for `sendFinished` just because some old code was trying to be clever (and thus hard to understand). In fact, the code confused me somewhat, until I realized that `self.retries` is the negative of the number of remaining retries.

Thanks for working on this. Please merge after changing the documentation of `sendFinished` to refer to `bool`, `True` and `False` and *new* code to use those values.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kaizhang set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_26"></a>|kaizhang set status to `closed`|
|-|-|

(In [39157]) Merge sendmail-deferred-cancellation-6572

Author: kaizhang
Reviewer: tom.prince
Fixes: [#6572](https://github.com/twisted/twisted/issues/6572)

Now `twisted.mail.smtp.sendmail` returns a cancellable Deferred. When a user cancels the Deferred, the `twisted.mail.smtp.SMTPSenderFactory` will stop retrying and disconnect the connection immediately.
